### DISCUSSION TITLE
Is it currently possible to read touch sensitive pins in esp32-s2?

### DISCUSSION THREAD

**Discussion by intGus** [Moddable-OpenSource/moddable#1263] [Q&A] [answered]
I have a lolin s2 mini, I read values from a touch sensitive  pin using micropython. I can see the values going up and down when I touch the pin directly in the board. I'm trying to do something similar with moddable but failed. I saw this module called touchpad and thought it would be what I need but I get this error:

`/home/user/.espressif/tools/xtensa-esp32s2-elf/esp-12.2.0_20230208/xtensa-esp32s2-elf/bin/../lib/gcc/xtensa-esp32s2-elf/12.2.0/../../../../xtensa-esp32s2-elf/bin/ld: xs_esp32s2.a(mc.xs.c.o):(.rodata.gxBuilders8+0xc): undefined reference to 'xs_touchpad'`

Here is my code:

```javascript
import TouchPad from "touchpad"

const Digital = device.io.Digital;

const led = new Digital({
  pin: 15,
  mode: Digital.Output,
});
led.write(0);

touch = new TouchPad({
  pin: 7
});
baseValue = touch.read()
```

Just for reference, the micropython code to set and read the touch sensitive pin is 
```python
touch_pin = TouchPad(Pin(3, mode=Pin.IN))
x = touch_pin.read()
```

**Comment by phoddie** [comment]
I should be possible using the TouchPad. However, it appeared that the TouchPad driver build was broken in the ESP-IDF 5 upgrade. I think @mkellner can help.

**Comment by mkellner** [comment, marked as answer]
Adding a few include paths to `tools/mcconfig/make.esp32.mk` allows the touch component to be built. If you have additional problems, please include the `manifest.json` file for the app so we can ensure you have the right includes and configuration.

```
diff --git a/tools/mcconfig/make.esp32.mk b/tools/mcconfig/make.esp32.mk
index ffc0dcfff..355ee796d 100644
--- a/tools/mcconfig/make.esp32.mk
+++ b/tools/mcconfig/make.esp32.mk
@@ -227,6 +227,8 @@ INC_DIRS = \
        $(IDF_PATH)/components/spi_flash/include \
        $(IDF_PATH)/components/tcpip_adapter/include \
        $(IDF_PATH)/components/tcpip_adapter \
+       $(IDF_PATH)/components/driver/touch_sensor/include \
+       $(IDF_PATH)/components/driver/touch_sensor/$(ESP32_SUBCLASS)/include \
        $(IDF_PATH)/components/tinyusb/additions/include
 
 XS_OBJ = \
```

This change (and the related `nmake.esp32.mk` change) will be included in the next open source drop.

**Reply by intGus** [reply]
Well, it worked! however I'm using kaluga as the target, but as soon as I touch the pins, I get the response I wanted that for now is just turning an LED on. I'll need to go through the kaluma setup-target and all related files to bring that into the s2 mini so I don't need to fake it.

Now my only other problem that I hope is not that difficult to solve is I can't get the trace through xsbug or console, or even serial messages to work, I can see this lines, then nothing:

```
<?xs-00000000?>

<?xs.3FFFE794?>
<xsbug><login name="main" value="XS 14.2.0 32-bit LE 16-bit ID" flags="E"/></xsbug>

<?xs-3FFFE794?>
```
I'm debugging using the LED which is a pain in the rear, if you can give me some ideas I'd appreciate it. Thanks!

**Comment by phoddie** [comment]
> Now my only other problem that I hope is not that difficult to solve is I can't get the trace through xsbug or console, or even serial messages to work, I can see this lines, then nothing:...

The Lins you quote are the debugging protocol starting up on the ESP32. It sends the login message and waits briefly for a response from xsbug. It isn't getting one, so it continues running without debugging.

Normally you don't see those lines because `serial2xsbug` bridges between serial and xsbug when you build with `mcconfig -d -m -p esp32/kaluga`. Are you not running `serial2xsbug` somehow?

If the receive pin on your board wasn't properly connected to your computer, that would give this behavior too. But that's a rare mistake.

As an alternative, you can have traces and no debugging, using an instrumented build. `mcconfig -I -m -p esp32/kaluga`.

**Reply by intGus** [reply]
It's working! I'm using xs-dev, and I wasn't able some of the errors I encountered after flashing the board. Following your advice, when I ran mcconfig, I noticed that the bash script waitForNewSerialLinux failed because findUSBLinux is not recognized as a command (presumably not in my PATH variable). To address this, I modified the two lines in the waitForNewSerialLinux script where it calls findUSBLinux to include the full path to the script. Now, I can see the logs in xsbug!

However, I'm still unable to view the output log in the terminal. I switched my node version, installed npm in the xsbug-log folder, but still, nothing appears when using -dl or -l. If I use -i I get an error from esp-idf-monitor. Nevertheless, I'm happy that I can at least see some progress now!

I forgot to mention these lines:

```
<?xs-00000000?>

<?xs.3FFFE794?>
<xsbug><login name="main" value="XS 14.2.0 32-bit LE 16-bit ID" flags="E"/></xsbug>

<?xs-3FFFE794?>
```

Were in my unsuccessful attempt to see something printed, using a serial monitor, not in xsbug.

Thank you so much!

**Reply by phoddie** [reply]
> It's working!

Congratulations! Nice work.

Since xsbug works, there's no fundamental reason xsbug-log shouldn't also work. They both communicate with the ESP32 via serial2xsbug using the same debugging protocol.

Do you see anything in the output that suggests what might not be working? When I run `mcconfig -dl -m -p esp32` on macOS, the relevant bit is here:

```
Leaving...
Hard resetting via RTS pin...
Done
#xsbug-log listening on port 5002. ^C to exit.
#xsbug-log connected to "main"
```

Just in case, please check that xsbug-log is installed correctly:

```
cd $MODDABLE/tools/xsbug
node xsbug-log.js
```

That should output something like:

```
#xsbug-log: no command line arguments. Waiting for connection.
#xsbug-log listening on port 5002. ^C to exit.
```

> presumably not in my PATH variable

If that's not set-up, I could imagine there will be other problems. Probably best to get that in place.

**Reply by intGus** [reply]
Sorry for the late response, I can make it work if I execute xsbug-log and then serial2xsbug, not if I do it thru the mcconfig, it is the same issue that I need to add everything to my PATH, but I can find why that happened. Thanks for the help!

**Reply by phoddie** [reply]
Excellent! 

Is there anything from your experience that we can learn from to help others? If there's something we can do in the docs or tools or xs-dev, please let us know. Thank you.
